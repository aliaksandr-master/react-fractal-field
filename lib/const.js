
export const TYPE = {
  OBJECT: 'object',
  ARRAY: 'array',
  SIMPLE: 'simple'
};
