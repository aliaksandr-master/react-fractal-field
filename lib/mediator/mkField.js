export const mkMeta = () => ({
  valid: true,
  touched: false,

  active: false,
  activated: false,

  submitted: false,
  submitting: false,
  submitFailed: false,
  submitSuccess: false,
  submittedTimes: 0,

  error: null,
  errorsIn: [],
  hasException: false,
  hasFormatException: false,
  hasNormalizeException: false,
  hasValidateException: false
});

export default ({
  id,
  complexValueType = null,
  parentID = null,
  autoClean = false,
  shareMeta = true,
  shareValue = true,
  value = undefined,
  name = null
}) => ({
  // permanent
  id,
  parentID,
  autoClean,
  shareMeta,
  shareValue,
  complexValueType, // if not null component might have named children

  // free transform
  name,

  value,

  meta: mkMeta(),

  children: [],
  version: 0,
  metaVersion: 0,
  valueVersion: 0
});
