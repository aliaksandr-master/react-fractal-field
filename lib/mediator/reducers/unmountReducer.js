import omit from 'lodash/omit';
import recalcValidity from '../recalcValidity';
import recalcValue from '../recalcValue';



export default () => (state, { fieldID }) => {
  const field = state[fieldID];
  const parentID = field.parentID;

  state = omit(state, fieldID);

  if (parentID) {
    const parent = state[field.parentID];

    if (parent) {
      state = {
        ...state,
        [parentID]: {
          ...parent,
          children: parent.children.filter((id) => id !== fieldID)
        }
      };
    }
  }

  state = recalcValidity(state, parentID, fieldID);

  state = recalcValue(state, parentID, field.id);

  return state;
};
