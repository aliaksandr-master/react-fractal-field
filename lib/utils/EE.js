import PrimitiveEE from './PrimitiveEE';



const EE = () => {
  let listeners = {};
  let destroyEE = PrimitiveEE();
  let everyEE = PrimitiveEE();
  let destroyed = false;

  const ee = {
    onEveryMessage (callback) {
      if (destroyed) {
        return () => {};
      }

      return everyEE.on(callback);
    },

    postMessage (eventName, ...args) {
      if (destroyed) {
        return;
      }

      if (!listeners.hasOwnProperty(eventName)) {
        return;
      }

      listeners[eventName].emit(...args);

      everyEE.emit(eventName, ...args);
    },

    onMessage (eventName, callback) {
      if (destroyed) {
        return () => {};
      }

      if (!listeners.hasOwnProperty(eventName)) {
        const listener = listeners[eventName] = PrimitiveEE();

        ee.onDestroy(() => {
          listener.cleanup();
          listeners[eventName] = null;
        });
      }

      return listeners[eventName].on(callback);
    },

    isDestroyed () {
      return destroyed;
    },

    destroy () {
      if (destroyed) {
        return;
      }

      destroyed = true;
      everyEE.cleanup();
      everyEE = null;
      destroyEE.emit();
      destroyEE.cleanup();
      destroyEE = null;
      listeners = null;
    },

    onDestroy (callback) {
      if (destroyed) {
        return () => {};
      }

      return destroyEE.on(callback);
    },

    childEE () {
      if (destroyed) {
        throw new Error('EE has already destroyed. you couldn\'t create any child');
      }

      const childEE = EE();

      const off = ee.onDestroy(() => {
        childEE.destroy();
      });

      childEE.onDestroy(off);

      childEE.postMessage = ee.postMessage;

      childEE.onMessage = (...args) => {
        const off = ee.onMessage(...args);

        childEE.onDestroy(off);

        return off;
      };

      return childEE;
    },

    connect (parentEE) {
      if (destroyed) {
        throw new Error('EE has already destroyed. you couldn\'t connect any child');
      }

      const off = parentEE.onDestroy(() => {
        ee.destroy();
      });

      ee.onDestroy(off);

      return ee;
    }
  };

  return ee;
};


export default EE;
