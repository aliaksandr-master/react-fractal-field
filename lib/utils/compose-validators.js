import isFunction from 'lodash/isFunction';
import isString from 'lodash/isString';
import isNil from 'lodash/isNil';



const allIsOk = () => undefined;

export default (...validators) => {
  validators = validators.filter((validator, index) => {
    if (isNil(validator)) {
      return false;
    }

    const validatorIsFunction = isFunction(validator);

    if (process.env.NODE_ENV !== 'production') {
      if (!validatorIsFunction) {
        throw new TypeError(`validator[${index}] is not a function`);
      }
    }

    return validatorIsFunction;
  });

  const len = validators.length;

  if (!len) {
    return allIsOk;
  }

  if (len === 1) {
    return validators[0];
  }

  return (value) => {
    for (let i = 0; i < len; i++) { // eslint-disable-line fp/no-loops
      const message = validators[i](value);

      if (!isNil(message)) {
        if (process.env.NODE_ENV !== 'production') {
          if (!isString(message)) {
            throw new TypeError('validator message must be string');
          }
        }

        return message;
      }
    }

    return undefined;
  };
};
