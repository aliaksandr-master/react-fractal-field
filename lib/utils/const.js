
export const TYPE = {
  OBJECT: 'object',
  ARRAY: 'array',
  SIMPLE: 'simple'
};

export const SHARE = {
  ALL: 'all',
  NONE: 'none',
  META: 'meta',
  VALUE: 'value'
};
