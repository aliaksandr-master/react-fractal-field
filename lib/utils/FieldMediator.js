import omit from 'lodash/omit';
import isEqual from 'lodash/isEqual';
import isNil from 'lodash/isNil';
import wrapDestroyed from '../utils/wrapDestroyed';
import Store from '../utils/Store';
import PrimitiveEE from '../utils/PrimitiveEE';
import { logError, logDispatch } from '../utils/log';
import reducer, { MOUNT, UNMOUNT, SUBMIT, SUBMIT_OK, SUBMIT_FAILED, SUBMIT_START, VALIDATE_EXCEPTION, NORMALIZE_EXCEPTION, FORMAT_EXCEPTION, FOCUS, TOUCH, CHANGE_VALIDITY, BLUR, CHANGE_VALUE, RENAME } from './FieldMediator.reducer';



export default class FractalMediator {

  constructor (id, parent = null, { debug = false } = {}) {
    this.id = id;
    this._parent = parent;
    this._destroyed = false;
    this._debug = (parent ? parent._debug : debug) || debug;
    this._store = parent ? parent._store : new Store(reducer);

    this.destroy = wrapDestroyed(this.destroy);
    this.onChange = wrapDestroyed(this.onChange, () => () => {});
    this.transaction = wrapDestroyed(this.transaction, () => () => {});
  }

  dehydrateNames (ids) {
    const fullState = this._store.getState();

    return ids.map((id) => fullState.fields[id].name);
  }

  destroy () {
    this._dispatch(UNMOUNT);
    this._destroyed = true;
  }

  transaction () {
    return this._store.transaction();
  }

  version () {
    return this._store.version;
  }

  mount (params = {}) {
    const parentID = this._parent ? this._parent.id : null;

    this._dispatch(MOUNT, { parentID, ...params });
  }

  getState () {
    return this._store.getState().fields[this.id];
  }

  _dispatch (type, payload) {
    const flush = this.transaction();

    const action = { type, fieldID: this.id, payload };

    const endDispatch = logDispatch(this._debug, this._store, action);

    this._store.dispatch(action);

    endDispatch();

    flush();
  }

  hasNamedChildren () {
    const fullState = this._store.getState();

    return fullState.tree[this.id].some((id) => !isNil(fullState.fields[id].name));
  }

  focus () {
    this._dispatch(FOCUS);
  }

  blur () {
    this._dispatch(BLUR);
  }

  submitStart () {
    this._dispatch(SUBMIT_START);
  }

  submit () {
    this._dispatch(SUBMIT);
  }

  submitDone () {
    this._dispatch(SUBMIT_OK);
  }

  submitFailed () {
    this._dispatch(SUBMIT_FAILED);
  }

  sysNormalizeError (err) {
    this._dispatch(NORMALIZE_EXCEPTION, err);
    if (err) {
      logError(err, 'normalize');
    }
  }

  sysFormatError (err) {
    this._dispatch(FORMAT_EXCEPTION, err);
    if (err) {
      logError(err, 'format');
    }
  }

  sysValidateError (err) {
    this._dispatch(VALIDATE_EXCEPTION, err);
    if (err) {
      logError(err, 'validate');
    }
  }

  changeValidity (error) {
    this._dispatch(CHANGE_VALIDITY, error || null);
  }

  onChange (listener) {
    return this._store.onChange(listener);
  }

  rename (name) {
    this._dispatch(RENAME, name);
  }

  changeValue (value) {
    const flush = this.transaction();

    this._dispatch(CHANGE_VALUE, value);
    this._dispatch(TOUCH);

    flush();
  }
}
