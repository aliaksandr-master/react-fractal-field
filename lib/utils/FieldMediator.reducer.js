import isEqual from 'lodash/isEqual';
import omit from 'lodash/omit';
import { TYPE, SHARE } from '../utils/const';
import reducer from './reducer';



export const MOUNT = 'MOUNT';
export const UNMOUNT = 'UNMOUNT';
export const RENAME = 'RENAME';
export const CHANGE_VALUE = 'CHANGE_VALUE';
export const CHANGE_VALIDITY = 'CHANGE_VALIDITY';
export const FORMAT_EXCEPTION = 'FORMAT_EXCEPTION';
export const NORMALIZE_EXCEPTION = 'NORMALIZE_EXCEPTION';
export const VALIDATE_EXCEPTION = 'VALIDATE_EXCEPTION';
export const SUBMIT_START = 'SUBMIT_START';
export const SUBMIT_FAILED = 'SUBMIT_FAILED';
export const SUBMIT_OK = 'SUBMIT_OK';
export const FOCUS = 'FOCUS';
export const BLUR = 'BLUR';
export const SUBMIT = 'SUBMIT';
export const TOUCH = 'TOUCH';



const hasShareLevel = (shareLevel) => {
  if (shareLevel === SHARE.ALL) {
    return ({ share }) => share !== SHARE.NONE;
  }

  if (shareLevel === SHARE.META) {
    return ({ share }) => share !== SHARE.NONE && share !== SHARE.VALUE;
  }

  if (shareLevel === SHARE.VALUE) {
    return ({ share }) => share !== SHARE.NONE && share !== SHARE.META;
  }

  return () => false;
};



const recursiveRecalc = (func, { skipIsolated = false, check = () => true } = {}) => {
  const apply = (state, fieldID) => {
    const field = state.fields[fieldID];

    if (!field) {
      return state;
    }

    const newState = func(state, field);

    if (newState === state) {
      return state;
    }

    if (!check(field)) {
      return state;
    }

    return apply(newState, field.parentID);
  };

  return apply;
};



const recalcValidity = recursiveRecalc((state, field) => {
  let valid = !(field.error || field.hasException);

  if (!valid && !field.valid) {
    return state;
  }

  let errorsIn = [];

  if (valid) {
    errorsIn = state.tree[field.id].filter((childID) => !state.fields[childID].valid);

    valid = !errorsIn.length;
  }

  if (valid === field.valid && isEqual(errorsIn, field.errorsIn)) {
    return state;
  }

  return {
    ...state,
    fields: {
      ...state.fields,
      [field.id]: {
        ...field,
        valid,
        errorsIn
      }
    }
  };
}, { check: hasShareLevel(SHARE.META) });

const recalcActivity = recursiveRecalc((state, field) => {
  const someChildIsActive = state.tree[field.id].some((childID) => state.fields[childID].active);

  if (someChildIsActive === field.active) {
    return state;
  }

  return {
    ...state,
    fields: {
      ...state.fields,
      [field.id]: {
        ...field,
        active: someChildIsActive,
        activated: true
      }
    }
  };
}, { check: hasShareLevel(SHARE.META) });

const recalcValue = recursiveRecalc((state, field) => {
  const children = state.tree[field.id];

  const value = children.reduce((value, childID) => {
    const child = state.fields[childID];

    if (!child.name) {
      return value;
    }

    value[child.name] = child.value;

    return value;
  }, field.type === TYPE.ARRAY ? [] : {});

  if (isEqual(field.value, value)) {
    return state;
  }

  return {
    ...state,
    fields: {
      ...state.fields,
      [field.id]: {
        ...field,
        value
      }
    }
  };
}, { check: hasShareLevel(SHARE.VALUE) });



// REDUCERS

const exceptionPropsMap = [ 'hasFormatException', 'hasNormalizeException', 'hasValidateException' ]
  .reduce((result, value, _i, array) => ({
    ...result,
    [value]: array.filter((v) => v !== value)
  }), {});

const exceptionReducer = (propName) => (state, { fieldID, payload: exception }) => {
  const field = state.fields[fieldID];

  exception = Boolean(exception);

  if (field[propName] === exception) {
    return state;
  }

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        valid: field.valid && !exception,
        hasException: exception || exceptionPropsMap[propName].some(Boolean),
        [propName]: exception
      }
    }
  };

  return recalcValidity(state, field.parentID);
};


const renameReducer = () => (state, { fieldID, payload: name = null }) => {
  const field = state.fields[fieldID];

  if (field.name === name) {
    return state;
  }

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: { ...field, name }
    }
  };

  return recalcValue(state, field.parentID);
};

const changeValidityReducer = () => (state, { fieldID, payload: error }) => {
  const field = state.fields[fieldID];

  if (isEqual(field.error, error)) {
    return state;
  }

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: { ...field, error }
    }
  };

  const newState = recalcValidity(state, field.id);

  if (newState === state) {
    return state;
  }

  return recalcValue(newState, field.parentID);
};

const mountReducer = () => (state, { fieldID, payload: { parentID, name = null, ...params } }) => {
  const field = {
    id: fieldID,
    name,
    valid: true,
    share: SHARE.ALL,
    value: undefined,
    touched: false,

    active: false,
    activated: false,

    submitted: 0,
    submitting: false,
    submitSuccess: false,
    submitFailed: false,

    error: null,
    errorsIn: [],
    hasException: false,
    hasFormatException: false,
    hasNormalizeException: false,
    hasValidateException: false,

    parentID,
    ...params
  };

  state = {
    ...state,
    tree: {
      ...state.tree,
      ...(parentID ? { [parentID]: state.tree[parentID] ? [ ...state.tree[parentID], fieldID ] : [ fieldID ] } : {}),
      [fieldID]: []
    },
    fields: {
      ...state.fields,
      [fieldID]: field
    }
  };

  return recalcValue(state, parentID);
};

const unmountReducer = () => (state, { fieldID }) => {
  const field = state.fields[fieldID];
  const parentID = field.parentID;

  state = {
    ...state,
    fields: omit(state.fields, fieldID),
    tree: {
      ...state.tree,
      [parentID]: state.tree[parentID].filter((id) => id !== fieldID)
    }
  };

  state = recalcValidity(state, parentID);

  return recalcValue(state, parentID);
};

const submitStartReducer = () => (state, { fieldID }) => {
  const field = state.fields[fieldID];

  if (field.submitting) {
    return state;
  }

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        submitting: true
      }
    }
  };

  return state;
};

const submitReducer = () => (state, { fieldID }) => {
  const field = state.fields[fieldID];

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        submitted: field.submitted + 1,
        submitting: false
      }
    }
  };

  return state;
};

const submitOkStartReducer = () => (state, { fieldID }) => {
  const field = state.fields[fieldID];

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        submitting: false,
        submitSuccess: true,
        submitFailed: false,
        submitted: field.submitted + 1
      }
    }
  };

  return state;
};

const submitFailedReducer = () => (state, { fieldID }) => {
  const field = state.fields[fieldID];

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        submitting: false,
        submitSuccess: false,
        submitFailed: true,
        submitted: field.submitted + 1
      }
    }
  };

  return state;
};

const focusReducer = () => (state, { fieldID }) => {
  const field = state.fields[fieldID];

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        active: true,
        activated: true
      }
    }
  };

  return recalcActivity(state, field.parentID);
};

const blurReducer = () => (state, { fieldID }) => {
  const field = state.fields[fieldID];

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        active: false
      }
    }
  };

  return recalcActivity(state, field.parentID);
};

const touchReducer = () => (state, { fieldID }) => {
  const field = state.fields[fieldID];

  if (field.touched) {
    return state;
  }

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        touched: true
      }
    }
  };

  return state;
};

const changeValueReducer = () => (state, { fieldID, payload: value }) => {
  const field = state.fields[fieldID];

  if (isEqual(field.value, value)) {
    return state;
  }

  state = {
    ...state,
    fields: {
      ...state.fields,
      [fieldID]: {
        ...field,
        value
      }
    }
  };

  return recalcValue(state, field.parentID);
};


export default reducer({
  fields: {},
  tree: {}
}, {
  [MOUNT]: mountReducer(),
  [UNMOUNT]: unmountReducer(),
  [SUBMIT]: submitReducer(),
  [SUBMIT_START]: submitStartReducer(),
  [SUBMIT_OK]: submitOkStartReducer(),
  [SUBMIT_FAILED]: submitFailedReducer(),
  [FOCUS]: focusReducer(),
  [BLUR]: blurReducer(),
  [TOUCH]: touchReducer(),
  [CHANGE_VALUE]: changeValueReducer(),
  [CHANGE_VALIDITY]: changeValidityReducer(),
  [RENAME]: renameReducer(),
  [FORMAT_EXCEPTION]: exceptionReducer('hasFormatException'),
  [NORMALIZE_EXCEPTION]: exceptionReducer('hasNormalizeException'),
  [VALIDATE_EXCEPTION]: exceptionReducer('hasValidateException')
});
