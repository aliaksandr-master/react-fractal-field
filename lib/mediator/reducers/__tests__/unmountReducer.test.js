/*eslint-env jest*/
import unmountReducer from '../unmountReducer';
import mkField from '../../mkField';
import { TYPE } from '../../../utils/const';


test('unmountReducer', () => {
  expect(unmountReducer()({}, { type: 'some', fieldID: 123 })).toMatchSnapshot();

  expect(unmountReducer()({ [123]: mkField({ id: 123 }) }, { type: 'some', fieldID: 123 })).toMatchSnapshot();

  const parent = mkField({ id: 123, complexValueType: TYPE.OBJECT, value: { [123]: 234 } });

  parent.children = [ 234 ];

  expect(unmountReducer()({ [123]: parent, [234]: mkField({ id: 234, parentID: 123, autoClean: true, name: 'some' }) }, { type: 'some', fieldID: 234 })).toMatchSnapshot();
});
