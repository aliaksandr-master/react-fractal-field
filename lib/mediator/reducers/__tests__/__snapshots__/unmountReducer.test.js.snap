// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unmountReducer 1`] = `Object {}`;

exports[`unmountReducer 2`] = `Object {}`;

exports[`unmountReducer 3`] = `
Object {
  "123": Object {
    "autoClean": false,
    "children": Array [],
    "complexValueType": "object",
    "id": 123,
    "meta": Object {
      "activated": false,
      "active": false,
      "error": null,
      "errorsIn": Array [],
      "hasException": false,
      "hasFormatException": false,
      "hasNormalizeException": false,
      "hasValidateException": false,
      "submitFailed": false,
      "submitSuccess": false,
      "submitted": false,
      "submittedTimes": 0,
      "submitting": false,
      "touched": false,
      "valid": true,
    },
    "metaVersion": 0,
    "name": null,
    "parentID": null,
    "shareMeta": true,
    "shareValue": true,
    "value": Object {
      "123": 234,
    },
    "valueVersion": 0,
    "version": 1,
  },
}
`;
